[SPARK-45856] Move ArtifactManager from Spark Connect into SparkSession (sql/core)

vicennial opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

The significant changes in this PR include:

- `SparkConnectArtifactManager` is renamed to `ArtifactManager` and moved out of Spark Connect and into `sql/core` (available in `SparkSession` through `SessionState`) along with all corresponding tests and confs.
 - While `ArtifactManager` in part of SparkSession, we keep the legacy behaviour for non-connect spark while utilising the ArtifactManager in connect pathways
    - This is done by exposing a new method `withResources` in the artifact manager that sets the context class loader (for driver-side operations) and propagates the `JobArtifactState` such that the resources reach the executor.
    - Spark Connect pathways utilise this method through the `SessionHolder#withActive`
    - When `withResources` is not used, neither the custom context classloader nor the `JobArtifactState` is propagated and hence, non Spark Connect pathways remain with legacy behaviour.  

### Why are the changes needed?
The `ArtifactManager` that currently lies in the connect package can be moved into the wider sql/core package (e.g SparkSession) to expand the scope. This is possible because the `ArtifactManager` is tied solely to the `SparkSession#sessionUUID` and hence can be cleanly detached from Spark Connect and be made generally available.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No. Existing behaviour is kept intact for both non-connect and connect spark.


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Existing tests.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No.

vicennial committed:
[SPARK-45856] Move ArtifactManager from Spark Connect into SparkSession (sql/core)

vicennial committed:
Merge branch 'master' into SPARK-45856

vicennial committed:
mima exclusions

vicennial committed:
fix lint + javadoc

vicennial committed:
mima + javadoc fixes

vicennial commented:
PTAL @hvanhovell @HyukjinKwon

vicennial committed:
Merge branch 'master' into SPARK-45856

vicennial committed:
fix merge conflict

hvanhovell reviewed (approved):
LGTM

dongjoon-hyun reviewed (commented):
According to the JIRA, this is only for Apache Spark 4.0.0, right?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -243,6 +244,16 @@ class SparkSession private(
+   * Returns an `ArtifactManager` that supports adding, managing and using session-scoped artifacts
+   * (jars, classfiles, etc).
+   *
+   * @since 3.5.1
```
This should be 4.0.0 because this PR is for Apache Spark 4.0.0, @vicennial .

dongjoon-hyun reviewed (changes requested):
We still need to support configurations in a way because we didn't deprecate yet in the Apache Spark community.
```
spark.connect.copyFromLocalToFs.allowDestLocal
```

vicennial committed:
update ver + support deprecated config

dongjoon-hyun commented:
Could you re-trigger the failed pipelines?

vicennial commented:
@dongjoon-hyun Thank you for the review! 

I've updated the version and for the deprecated conf `spark.connect.copyFromLocalToFs.allowDestLocal`, I've added it to the `deprecatedSQLConfigs` list so that it would generate a warning if used. Currently, both configs would work but if the new config is set (currently optional), it would override the deprecated config. Eventually, we should remove the deprecated conf and convert the new config from optional to have a default value.

vicennial commented:
Hmm, JavaDocGeneration is failing in 2 tests and from the logs, its not clear why... 
I will merge master and see if that helps

vicennial committed:
Merge branch 'master' into SPARK-45856

vicennial commented:
@dongjoon-hyun The CI is green now :)

HyukjinKwon commented:
Merged to master.

fhalde commented:
Hi, we are super interested in having the isolated classloader per spark session ability for our usecase. i believe this today is only achievable if jobs are run from a connect client. we want to avoid using connect client

but with this pr merged, it should be possible to have isolated classloaders per spark session right without any use of spark connect?

our use-case involves starting a spark driver and dynamically loading & adding jars and running transformations present within the jars. without isolated classloaders per session, on the executor side we would risk classpath conflict

@HyukjinKwon if your pr addresses my concern, i can back port it to 3.5.0 in my fork

vicennial commented:
@fhalde Yes, with this PR, it would be possible to have isolated classloaders per spark session on the executors without going through Spark Connect. 

The `withResources` method should be used to wrap all executions (like in Spark Connect [here](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala#L245)) and note that, all artifacts (i.e Jars, classfiles) would need to be added through the `ArtifactManager` (directly adding these to `SparkContext` would not work), refer to Spark Connect's `AddArtifactsHandler` to see how we use the [API](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala#L90).

LuciferYang commented on project/MimaExcludes.scala:
```diff
@@ -90,6 +90,16 @@ object MimaExcludes {
     ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.streaming.StreamingQueryException"),
 
+    // SPARK-45856: Move ArtifactManager from Spark Connect into SparkSession (sql/core)
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.storage.CacheId.apply"),
```
@vicennial I would like to reconfirm, the `ProblemFilters` added by SPARK-45856 will never need to undergo a mima check in versions after Spark 4.0, is that correct? I found that it has been placed in `defaultExcludes`.
